libpromises/eval_context.c

```diff
@@ -2111,6 +2116,17 @@ static Variable *VariableResolve(const EvalContext *ctx, const VarRef *ref)
         }
     }
 
+    // Try to qualify "this." variable to "current_bundle."
+    const Bundle *last_bundle = EvalContextStackCurrentBundle(ctx);
+    if (last_bundle && strcmp(ref->scope, SpecialScopeToString(SPECIAL_SCOPE_THIS)) == 0)
```

Comment:
Basically the fix is to modify this to also check if scope is SPECIAL_SCOPE_BODY. But I've made so many modifications until I found it, and then I also wanted to purge the recursion, that I submitted the whole thing in another PR: https://github.com/cfengine/core/pull/3024